### PR TITLE
ingest/ledgerbackend: Restart captive core when a new version of core is detected on disk

### PR DESCRIPTION
**Pull request by tamirms**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [ ] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [ ] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [ ] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [ ] This PR adds tests for the most critical parts of the new functionality or fixes.
* [ ] I've updated any docs ([developer docs](https://www.stellar.org/developers/reference/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [ ] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What

Close https://github.com/stellar/go/issues/3602

* Add a ticker to stellarCoreRunner which will execute every 10 seconds. The stellarCoreRunner context will interrupt the goroutine attached to the ticker so that if captive core shutsdown the ticker will also stop.

* Upon every tick, we will check to see if the stellar core binary has been modified. If it has been modified we log a warning message to announce we will be shutting down captive core and then we call stellarCoreRunner.close() to terminate the captive core instance. Once stellarCoreRunner.close() is called any pending operations (PrepareRange() or GetLedger()) should return immediately with an error and the ingestion state machine will handle the error by retrying.

* When the retry happens it will create a new instance of stellarCoreRunner which should use the newest stellar core binary.

### Why

See https://github.com/stellar/go/issues/3602 for motiviation.

### CODE REVIEW

**Review by bartekn — APPROVED** [review]

File: services/horizon/CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,7 @@ file. This project adheres to [Semantic Versioning](http://semver.org/).
 
 * Performance improvement: Captive Core now reuses bucket files whenever it finds existing ones in the corresponding `--captive-core-storage-path` (introduced in [v2.1.0](#v2.1.0) rather than generating a one-time temporary sub-directory ([](https://github.com/stellar/go/pull/XXXX)). **This feature requires Stellar-Core version 17.1 or later.**
```
Comment (Shaptic) [inline services/horizon/CHANGELOG.md:12]: Unrelated but could you fix the PR link here to reference #3670? Forgot to update it after the fact.

File: services/horizon/CHANGELOG.md [hunk 2]
```diff
@@ -11,6 +11,7 @@ file. This project adheres to [Semantic Versioning](http://semver.org/).
...
 
+* Horizon monitors captive-core on disk, and restarts captive-core if it detects a change (e.g. a more recent file timestamp for the captive-core binary). ([3687](https://github.com/stellar/go/pull/3687)).
```
Comment (Shaptic) [inline services/horizon/CHANGELOG.md:14]: It monitors the Stellar Core binary, right? Well, more specifically, the binary pointed to by `--stellar-core-binary-path`. Maybe something like:

```suggestion
* Horizon now monitors the Stellar Core binary on disk (pointed to by `--stellar-core-binary-path`/`STELLAR_CORE_BINARY_PATH`) and restarts its Captive Core subprocess if it detects changes (i.e a more recent file timestamp for the Stellar Core binary) ([3687](https://github.com/stellar/go/pull/3687)).
```

**Review by Shaptic — APPROVED** [review]
:+1:

File: ingest/ledgerbackend/stellar_core_runner.go [hunk 3]
```diff
@@ -249,14 +249,18 @@ func (r *stellarCoreRunner) catchup(from, to uint32) error {
 		return errors.Wrap(err, "error waiting for `stellar-core new-db` subprocess")
 	}
 
+	binaryWatcher, err := newFileWatcher(r)
+	if err != nil {
+		return errors.Wrap(err, "could not create captive core binary watcher")
+	}
+
```
[the remark this answers is not in the document]
Comment (Shaptic) [reply ingest/ledgerbackend/stellar_core_runner.go:256]: Though it seems like the only error case is where we can't `stat` the stellar-core binary, which means we probably failed way before this anyway...

**Comment by bartekn** [comment]
@tamirms please merge when ready.

**Pull request merged**